src/main/java/life/qbic/model/download/QbicDataDownloader.java

```diff
@@ -187,134 +193,185 @@ private void downloadDatasets(Sample analyte, List<DataSet> dataSets,
       LOG.info(
           "Downloading " + files.size() + " file" + (files.size() != 1 ? "s" : "") + " for dataset "
               + datasetSample + " (" + permID + ")");
-      downloadFiles(downloadRequest);
+      downloadResponse.merge(downloadFiles(downloadRequest));
     }
+    return downloadResponse;
   }
 
-  private void downloadFile(DataSetFile dataSetFile, Path prefix) throws DownloadException {
+  private FileDownloadResponse downloadFile(DataSetFile dataSetFile, Path prefix) throws DownloadException {
     for (IDataStoreServerApi dataStoreServer : dataStoreServers) {
       try {
-        downloadFileFromDataStoreServer(dataSetFile, prefix, dataStoreServer);
-        break;
+        return downloadFileFromDataStoreServer(dataSetFile, prefix, dataStoreServer);
       } catch (FileNotFoundException fileNotFoundException) {
         // log and try the next data store server
         LOG.debug("Download attempt failed on " + dataStoreServer, fileNotFoundException);
-        continue;
       }
     }
+    return FileDownloadResponse.failure();
   }
 
+  // note: DataSetFileDownloadReader closes the input stream after it finished reading it.
+  private static class AutoClosableDataSetFileDownloadReader extends DataSetFileDownloadReader implements AutoCloseable {
+    public AutoClosableDataSetFileDownloadReader(InputStream in) {
+      super(in);
+    }
+  }
+
+  public static class FileDownloadResponse {
+    private final boolean success;
+
+    private FileDownloadResponse(boolean success) {
+      this.success = success;
+    }
+
+    public static FileDownloadResponse success() {
+      return new FileDownloadResponse(true);
+    }
+
+    private static FileDownloadResponse failure() {
+      return new FileDownloadResponse(false);
+    }
+
+    public boolean isSuccess() {
+      return success;
+    }
+
+    public boolean isFailure() {
+      return !success;
+    }
```

Comment:
Suggestion: since you introduce this API already, might be worth to add some more expressive methods, like `whenSuccess(Function f)` and `whenSuccessOrElse(Function onSuccess, Function onFailure)`.

Reply:
As I need to accumulate the results, a consumer would not help me here. This is a possible extension in the case that we refactor the tool to a more pipeline-like structure.